detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt

```diff
@@ -0,0 +1,117 @@
+        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK_BROADCAST)] = ::onHalfBlockBroadcastPacket
+        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK_ENCRYPTED)] = ::onHalfBlockPacket
+        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK_BROADCAST_ENCRYPTED)] = ::onHalfBlockBroadcastPacket
+    }
+
+    fun sendTransaction(blockBuilder: BlockBuilder, peer: Peer?, encrypt: Boolean = false) {
+        println("sending block...")
+        val block = blockBuilder.sign()
+
+        if (peer != null) {
+            sendBlockToRecipient(peer, block, encrypt)
+        } else {
+            sendBlockBroadcast(block, encrypt)
+        }
+    }
+
+    private fun sendBlockToRecipient(peer: Peer, block: TrustChainBlock, encrypt: Boolean) {
+        val payload = HalfBlockPayload.fromHalfBlock(block)
+
+        val data = if (encrypt) {
+            serializePacket(MessageId.HALF_BLOCK_ENCRYPTED, payload, false, encrypt = true, recipient = peer)
+        } else {
+            serializePacket(MessageId.HALF_BLOCK, payload, false)
+        }
+
+        send(peer, data)
+    }
+
+    private fun sendBlockBroadcast(block: TrustChainBlock, encrypt: Boolean) {
+        val payload = HalfBlockBroadcastPayload.fromHalfBlock(block, ttl.toUInt())
+        val randomPeers = getPeers().random(broadcastFanOut)
+        for (randomPeer in randomPeers) {
+            val data = if (encrypt) {
+                serializePacket(MessageId.HALF_BLOCK_BROADCAST_ENCRYPTED, payload, false, encrypt = true, recipient = randomPeer)
+            } else {
+                serializePacket(MessageId.HALF_BLOCK_BROADCAST, payload, false)
+            }
+            send(randomPeer, data)
+        }
+    }
+
+    private fun onHalfBlockPacket(packet: Packet) {
+        println("half block packet received from: " + packet.source.toString())
+    }
+
+    private fun onHalfBlockBroadcastPacket(packet: Packet) {
+        println("half block packet received from broadcast from: " + packet.source.toString())
+    }
+
+    override fun onPacket(packet: Packet) {
+        val sourceAddress = packet.source
+        val data = packet.data
+
+        val probablePeer = network.getVerifiedByAddress(sourceAddress)
+        if (probablePeer != null) {
+            probablePeer.lastResponse = Date()
+        }
+
+        val packetPrefix = data.copyOfRange(0, prefix.size)
+        if (!packetPrefix.contentEquals(prefix)) {
+            // logger.debug("prefix not matching")
+            return
+        }
+
+        val msgId = data[prefix.size].toUByte().toInt()
+
+        val handler = messageHandlers[msgId]
+
+        if (handler != null) {
+            try {
+                handler(packet)
+            } catch (e: Exception) {
+                e.printStackTrace()
+            }
+        } else {
+            println("Received unknown message $msgId from $sourceAddress")
+        }
+    }
+
+    private fun msgIdFixer(msgid: Int): Int{
+        @Suppress("DEPRECATION")
```

Comment:
Not a good practice, there should be some other approach for the same action

Reply:
I agree. The problem was that we used `MessageId` flags (e.g: `HALF_BLOCK`, `HALF_BLOCK_ENCRYPTED` etc...) that were larger than a `Ubyte`. I will use smaller values as flags and I will remove the `msgIdFixer`.